### ISSUE TITLE
openstack_images_image_v2 indicates changes every time due to read-only computed property

### ISSUE THREAD

**Issue by jrandall** [terraform-providers/terraform-provider-openstack#301]
### Terraform Version
terraform 0.11.2, terraform-provider-openstack 1.3.0

### Affected Resource(s)
- openstack_images_image_v2

### Terraform Configuration Files
```hcl
resource "openstack_images_image_v2" "hgi-image" {
  name             = "${var.env}-${var.region}-${var.setup}-${var.image_name}"
  image_source_url = "${var.image_url_base}/${var.image_name}"
  container_format = "${var.image_container_format}"
  disk_format      = "${var.image_disk_format}"

  properties {
    user = "beastie"
  }
}
```

### Expected Behavior
Image property `user` should be set and on subsequent runs, nothing should be changed.

### Actual Behavior
Property is set, but resource attempts to change every time to remove a read-only property `direct_url`:
```
Terraform will perform the following actions:

  ~ module.hgi-openstack-image-hgi-base-freebsd11-4cb02ffa.openstack_images_image_v2.hgi-image
      properties.direct_url:      "rbd://049fc780-8998-45a8-be12-d3b8b6f30e69/zeta-images/803a17db-8023-460f-9ea0-8f778b0f66d5/snap" => ""
      properties.user:            "" => "beastie"
```

### Important Factoids
Our openstack environment has glance backed by Ceph, and the images v2 schema (see https://developer.openstack.org/api-ref/image/v2/index.html#image-schemas) includes a read only property called `direct_url` that represents the backend storage location of the image data. 

terraform-provider-openstack does not seem to take the schema into account when determining which properties need updating, and as a result it is not possible to set any properties without causing terraform to indicate a change on every invocation. 

terraform-provider-openstack could use the image schemas API to check which properties are read-only, or perhaps we could add a provider or resource configuration option that would provide a list of read-only image properties that should be ignored when updating images?

**Comment by jtopjian** [comment]
Thanks for reporting this.

> terraform-provider-openstack could use the image schemas API to check which properties are read-only, or perhaps we could add a provider or resource configuration option that would provide a list of read-only image properties that should be ignored when updating images?

Probably the latter, most likely in the form of an `all_properties` attribute that would list combined properties from the Terraform user and from other sources.

In order to support the former, the schema part of the Glance API would need added to Gophercloud first.

**Comment by jrandall** [comment]
I'm not sure I understand how the `all_properties` attribute would work. I had imagined something like an `computed_properties` or `ignore_properties` argument that would configure the resource to not manage those properties. 

Is the idea of `all_properties` the inverse of that - it would provide a full list of all the properties that the terraform-provider-openstack cares about managing (so I would leave `direct_url` off the list so that it wouldn't be managed)? I guess that would have the advantage of tending to work out of the box without extra configuration. On the other hand, it would probably still be good to provide all of the properties (not just those listed in `all_properties`) as read-only resource attributes.

**Comment by jtopjian** [comment]
The idea is to make `properties` write-only: a user can set properties but Terraform will never set values returned by Glance.

Instead, Terraform will set all properties returned by Glance in the `all_properties` attribute. This attribute would be computed and read-only. This way if an OpenStack environment sets arbitrary useful information about the image upon creation, the user can read it and possibly use it as input for other Terraform resources.

**Comment by jtopjian** [comment]
Oops - I apologize! "write-only" is a little confusing in this situation. It's "write-only" in the way that a user should only ever use it to set properties -- never to read properties.

If the user is expecting new properties upon image creation, they will then use the `all_properties` attribute.

This pattern is done in a few other places in this provider -- notably `openstack_compute_instance_v2.metadata` and `all_metadata`  attributes.

Let me know if that clarifies things :)

**Comment by jrandall** [comment]
Ok, yes, the reference to the existing functionality in `openstack_compute_instance_v2` was helpful. So the provider would populate the `properties` attribute with only those properties that have been set in the `properties` argument (and this is what is used for the comparison to detect whether a change is required). The `all_properties` attribute would include those properties and also all of the other properties returned by openstack.

That sounds like a perfect solution, thanks!

**Comment by jrandall** [comment]
I started to implement this, but it looks like gophercloud may be missing a patch method for replacing or deleting image properties. It has replace methods (but not delete) for tags, checksum, and name, but it seems to be missing any patching for generic properties. Is there a preferred way of doing this?

**Comment by jtopjian** [comment]
Unfortunately, there's a good amount of functionality missing from Gophercloud for the Glance API. 

You raise a good point, though: the existing `properties` attribute should probably have `ForceNew: true` set since it can't be updated.

However, unless I'm missing something, there shouldn't be a need to update properties in order to implement the `all_properties` feature. The bulk of the fix is changing this:

https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_images_image_v2.go#L291

to look like:

```go
d.Set("all_properties", img.Properties)
```

There will need to be an `all_properties` schema attribute added and the existing `properties` schema attribute will no longer be computed.

This is just off the top of my head, but that should almost be all of it.

**Comment by jtopjian** [comment]
@jrandall I wanted to check in and see if you needed further help with this? If you'd like, I could submit a PR for this.

**Comment by jtopjian** [comment]
@jrandall I have opened #409 which should resolve this issue for you. I've added support specifically for the `direct_url` property so that it is automatically added to the `properties` attribute. This means that after creation, `direct_url` will automatically be set and subsequent runs of Terraform should not try to remove it.

I've left some notes in #409 about the originating causes and interim fixes which have been made to ultimately get this working as intended.

Let me know if you would be able to test it and confirm if it resolves the behavior you're seeing.
